### ISSUE TITLE
Infinite loop in `locate_with_hint_fixed_core` for Constrained Triangulation

### ISSUE THREAD

**Issue by dylanowen** [Stoeoef/spade#98]
I don't believe the lines I'm inserting are intersecting (at least they pass `can_add_constraint`). However when running `locate_with_hint_fixed_core` it's possible to loop back to the start when trying to figure out where to add the new point.

I was able to trim down my error case to this test. It fails for both f32 and f64
```rust
#[test]
    fn test_infinite_loop() {
        let mut triangulation = ConstrainedDelaunayTriangulation::<Point2<_>>::default();

        let start = Point2::new(-21.296192169189453, 9.872323036193848);
        let edges = [
            (Point2::new(-20.926544189453125, 16.53529167175293),Point2::new(-27.772645950317383, 4.197676658630371), ),
            (Point2::new(-20.03745460510254, 12.93730354309082),Point2::new(-20.930097579956055, 11.93786907196045), ),
            (Point2::new(-15.576859474182129, 8.772907257080078),Point2::new(-22.373262405395508, 12.348699569702148), ),
            (Point2::new(-10.038422584533691, 5.663522243499756),Point2::new(-16.382625579833984, 9.09498119354248), ),
            (Point2::new(0.0, 0.0),Point2::new(-13.11422061920166, 7.30709171295166), ),
            (Point2::new(-19.230497360229492, -3.7645812034606934),Point2::new(-7.411926746368408, 3.486957311630249), ),
            (Point2::new(-25.072885513305664, -9.239323616027832),Point2::new(-19.462360382080078, -1.621320366859436), ),
            (Point2::new(-32.41080856323242, -13.72575855255127),Point2::new(-22.58626365661621, -2.076631784439087), ),
            (Point2::new(-32.41080856323242, -13.72575855255127),Point2::new(-25.57504653930664, -4.952820301055908), ),
            (Point2::new(-33.08932113647461, 0.31093916296958923),Point2::new(-25.955543518066406, 0.18878456950187683), ),
        ];

        for (p1, p2) in edges {
            let p1 = triangulation.insert(p1).unwrap();
            let p2 = triangulation.insert(p2).unwrap();
            assert!(triangulation.can_add_constraint(p1, p2));
            triangulation.add_constraint(p1, p2);
        }

        triangulation.insert(start).unwrap();
    }
```

Besides fixing this bug it might be a good idea to introduce some fail safes with a panic for all the `loop` expressions. Something like what the geo project is doing:
https://github.com/georust/geo/blob/47fadf1d507fd18a40921be28ee52e1ada6def04/geo/src/algorithm/triangulate_spade.rs#L380

**Comment by Stoeoef** [comment]
Interesting case - thanks for the report!

Indeed, the `locate` method is based on an assumption that doesn't seem to hold for CDT's (somewhat surprisingly).

It assumes that, when starting from an arbitrary edge `edge` with `edge.side_query(target_position).is_on_left_side()`, repeatedly taking finding an adjacent (= `edge.next()` and `edge.prev()`) edge will take it closer to the target face *as long as the target position remains on the left side* .

In the point sample that you have provided, the constraint edges "overlap" just in the right way to loop indefinitely around the target vertice (red dot).

See this graphic and the infinitely repeating list of edges below for some visual guidance!

I'm not entirely sure how to fix this in a robust way yet unfortunately but I am working on it!
Introducing some fail safe counters also seems like a good idea in this case.

![bug](https://github.com/Stoeoef/spade/assets/5731969/dd9c3c79-c05d-4e4c-9723-e4a8a9c9ce9b)

```text
{ index: 3 } -> { index: 5 }
{ index: 4 } -> { index: 5 }
{ index: 4 } -> { index: 7 }
{ index: 6 } -> { index: 7 }
{ index: 6 } -> { index: 9 }
{ index: 8 } -> { index: 9 }
{ index: 8 } -> { index: 11 }
{ index: 10 } -> { index: 11 }
{ index: 10 } -> { index: 13 }
{ index: 12 } -> { index: 13 }
{ index: 12 } -> { index: 15 }
{ index: 14 } -> { index: 15 }
{ index: 14 } -> { index: 16 }
{ index: 17 } -> { index: 16 }
{ index: 17 } -> { index: 18 }
{ index: 17 } -> { index: 1 }
{ index: 0 } -> { index: 1 }
{ index: 0 } -> { index: 5 }
{ index: 2 } -> { index: 5 }
{ index: 2 } -> { index: 3 }
{ index: 4 } -> { index: 3 }
{ index: 4 } -> { index: 5 }
{ index: 4 } -> { index: 7 }
{ index: 6 } -> { index: 7 }
{ index: 6 } -> { index: 9 }
{ index: 8 } -> { index: 9 }
{ index: 8 } -> { index: 11 }
{ index: 10 } -> { index: 11 }
{ index: 10 } -> { index: 13 }
{ index: 12 } -> { index: 13 }
{ index: 12 } -> { index: 15 }
{ index: 14 } -> { index: 15 }
{ index: 14 } -> { index: 16 }
{ index: 17 } -> { index: 16 }
{ index: 17 } -> { index: 18 }
{ index: 17 } -> { index: 1 }
{ index: 0 } -> { index: 1 }
{ index: 0 } -> { index: 5 }
{ index: 2 } -> { index: 5 }
{ index: 2 } -> { index: 3 }
{ index: 4 } -> { index: 3 }
{ index: 4 } -> { index: 5 }
{ index: 4 } -> { index: 7 }
{ index: 6 } -> { index: 7 }
{ index: 6 } -> { index: 9 }
{ index: 8 } -> { index: 9 }
{ index: 8 } -> { index: 11 }
{ index: 10 } -> { index: 11 }
{ index: 10 } -> { index: 13 }
{ index: 12 } -> { index: 13 }
{ index: 12 } -> { index: 15 }
```

**Comment by Stoeoef** [comment]
Fixed with version 2.5.1 (just released) - thanks again for the report!

The solution I went for is:
- Create some "nasa style" safety counters that prevent infinite loops from happening (thanks for the suggestion!)
- Change the walk strategy: Previously, the algorithm would not rotate around a single vertex systematically when attempting to get closer to the target position. It would sometimes increase the distance of the current edge towards the target position.
The change implemented in #99  now rotates around `edge.opposite_vertex` until it finds the angle segment between two out edge pairs that contains the target position. The edge spanned by this angle segment then becomes the new rotation edge.
 This should always terminate since this such a segment always exists (with the exception of some special cases like convex hull edges) and advancing into this segment always reduces the projection distance of the current edge to the target location.
